README.md
# leaflet-rastercoords

> Leaflet plugin for plain image map projection to display large images using tiles generated with gdal2tiles-leaflet

[![NPM version](https://badge.fury.io/js/leaflet-rastercoords.svg)](https://www.npmjs.com/package/leaflet-rastercoords)

See the plugin in action on <https://commenthol.github.io/leaflet-rastercoords/>.

## Usage

Process your "large" image with [gdal2tiles-leaflet][]

```js
// for use with browserify / webpack
var L = require('leaflet')
L.RasterCoords = require('leaflet-rastercoords')

var img = [
  3831,  // original width of image (here from `example/karta.jpg`)
  3101   // original height of image
]
// create the map
var map = L.map('map', {
  crs: L.CRS.Simple
})

// assign map and image dimensions
var rc = new L.RasterCoords(map, img)
// set max zoom Level (might be `x` if gdal2tiles was called with `-z 0-x` option)
map.setMaxZoom(rc.zoomLevel())
// all coordinates need to be unprojected using the `unproject` method
// set the view in the lower right edge of the image
map.setView(rc.unproject([img[0], img[1]]), 2)

// set markers on click events in the map
map.on('click', function (event) {
  // any position in leaflet needs to be projected to obtain the image coordinates
  var coords = rc.project(event.latlng)
  var marker = L.marker(rc.unproject(coords))
    .addTo(map)
  marker.bindPopup('[' + Math.floor(coords.x) + ',' + Math.floor(coords.y) + ']')
    .openPopup()
})

// the tile layer containing the image generated with `gdal2tiles --leaflet -p raster -w none <img> tiles`
L.tileLayer('./tiles/{z}/{x}/{y}.png', {
  noWrap: true,
  bounds: rc.getMaxBounds(),
  maxNativeZoom: rc.zoomLevel()
}).addTo(map)
```

## Example

The tiles in the example were generated using [gdal2tiles-leaflet][].
Take a look at [example/createtiles.sh](./example/createtiles.sh).

## License

Copyright (c) 2016- commenthol (MIT License)  
See [LICENSE][] for more info.

## References

<!-- !ref -->

* [example][example]
* [gdal2tiles-leaflet][gdal2tiles-leaflet]
* [LICENSE][LICENSE]

<!-- ref! -->

[LICENSE]: ./LICENSE
[gdal2tiles-leaflet]: https://github.com/commenthol/gdal2tiles-leaflet
[example]: https://commenthol.github.io/gdal2tiles-leaflet/test/index.html


CHANGELOG.md
## <small>1.0.6 (2023-09-25)</small>

* chore: bump dependencies ([1fff169](https://github.com/commenthol/leaflet-rastercoords/commit/1fff169))

## <small>1.0.5 (2022-11-02)</small>

* FIX: setting the max boundaries of map is optional now ([fd57cb6](https://github.com/commenthol/leaflet-rastercoords/commit/fd57cb6))


## <small>1.0.4 (2021-11-03)</small>

* add bounds to docs ([ed0a051](https://github.com/commenthol/leaflet-rastercoords/commit/ed0a051))
* add method to get bounds ([6543678](https://github.com/commenthol/leaflet-rastercoords/commit/6543678))
* remove trailing comma ([1f4ac11](https://github.com/commenthol/leaflet-rastercoords/commit/1f4ac11))
* fix: linter ([bdee88f](https://github.com/commenthol/leaflet-rastercoords/commit/bdee88f))
* chore: bump dependencies ([18f26ff](https://github.com/commenthol/leaflet-rastercoords/commit/18f26ff))
* chore: update example with circles ([618cae4](https://github.com/commenthol/leaflet-rastercoords/commit/618cae4))
* chore: update gh-pages ([1bd7b1c](https://github.com/commenthol/leaflet-rastercoords/commit/1bd7b1c))
* chore(example): fix unpkg version and featureGroup ([adcc57f](https://github.com/commenthol/leaflet-rastercoords/commit/adcc57f))



## <small>1.0.3 (2018-08-14)</small>

* add semicolon at end for bower ([cbb8bd9](https://github.com/commenthol/leaflet-rastercoords/commit/cbb8bd9))
* version ([fb9295e](https://github.com/commenthol/leaflet-rastercoords/commit/fb9295e))



## <small>1.0.2 (2017-02-07)</small>

* fix documentation ([7d16aaa](https://github.com/commenthol/leaflet-rastercoords/commit/7d16aaa))



## <small>1.0.1 (2016-10-26)</small>

* fix badge ([7cd1d4a](https://github.com/commenthol/leaflet-rastercoords/commit/7cd1d4a))
* fix usage with browserify/webpack ([412a767](https://github.com/commenthol/leaflet-rastercoords/commit/412a767))
* update .json files ([57d22b1](https://github.com/commenthol/leaflet-rastercoords/commit/57d22b1))
* update link ([01a4a4c](https://github.com/commenthol/leaflet-rastercoords/commit/01a4a4c))
* using cdn ([3eb73fc](https://github.com/commenthol/leaflet-rastercoords/commit/3eb73fc))



## 1.0.0 (2016-10-22)

* Initial commit ([f8d05db](https://github.com/commenthol/leaflet-rastercoords/commit/f8d05db))
* packaged ([d9f2ddf](https://github.com/commenthol/leaflet-rastercoords/commit/d9f2ddf))





rastercoords.js
/**
 * leaflet plugin for plain image map projection
 * @copyright 2016- commenthol
 * @license MIT
 */
/* globals define */
/* eslint no-var:off */

;(function (factory) {
  var L
  if (typeof define === 'function' && define.amd) {
    // AMD
    define(['leaflet'], factory)
  } else if (typeof module !== 'undefined') {
    // Node/CommonJS
    L = require('leaflet')
    module.exports = factory(L)
  } else {
    // Browser globals
    if (typeof window.L === 'undefined') {
      throw new Error('Leaflet must be loaded first')
    }
    factory(window.L)
  }
}(function (L) {
  /**
   * L.RasterCoords
   * @param {L.map} map - the map used
   * @param {Array} imgsize - [ width, height ] image dimensions
   * @param {Number} [tilesize] - tilesize in pixels. Default=256
   * @param {Boolean} setmaxbounds - automatically set map max bounds. Default=true
   */
  L.RasterCoords = function (map, imgsize, tilesize, setmaxbounds = true) {
    this.map = map
    this.width = imgsize[0]
    this.height = imgsize[1]
    this.tilesize = tilesize || 256
    this.zoom = this.zoomLevel()
    if (setmaxbounds && this.width && this.height) {
      this.setMaxBounds()
    }
  }

  L.RasterCoords.prototype = {
    /**
     * calculate accurate zoom level for the given image size
     */
    zoomLevel: function () {
      return Math.ceil(
        Math.log(
          Math.max(this.width, this.height) /
          this.tilesize
        ) / Math.log(2)
      )
    },
    /**
     * unproject `coords` to the raster coordinates used by the raster image projection
     * @param {Array} coords - [ x, y ]
     * @return {L.LatLng} - internal coordinates
     */
    unproject: function (coords) {
      return this.map.unproject(coords, this.zoom)
    },
    /**
     * project `coords` back to image coordinates
     * @param {Array} coords - [ x, y ]
     * @return {L.LatLng} - image coordinates
     */
    project: function (coords) {
      return this.map.project(coords, this.zoom)
    },
    /**
     * get the max bounds of the image
     */
    getMaxBounds: function () {
      var southWest = this.unproject([0, this.height])
      var northEast = this.unproject([this.width, 0])
      return new L.LatLngBounds(southWest, northEast)
    },
    /**
     * sets the max bounds on map
     */
    setMaxBounds: function () {
      var bounds = this.getMaxBounds()
      this.map.setMaxBounds(bounds)
    }
  }

  return L.RasterCoords
}))
; // eslint-disable-line semi


example/geojson.js
;(function (window) {
  // geoJson definitions for country
  window.countries = [{
    type: 'Feature',
    properties: {
      name: 'Iceland'
    },
    geometry: {
      type: 'Point',
      coordinates: [1258, 911]
    }
  }, {
    type: 'Feature',
    properties: {
      name: 'Ireland'
    },
    geometry: {
      type: 'Point',
      coordinates: [1324, 1580]
    }
  }, {
    type: 'Feature',
    properties: {
      name: 'England'
    },
    geometry: {
      type: 'Point',
      coordinates: [1498, 1662]
    }
  }, {
    type: 'Feature',
    properties: {
      name: 'France'
    },
    geometry: {
      type: 'Point',
      coordinates: [1608, 1918]
    }
  }, {
    type: 'Feature',
    properties: {
      name: 'Italia'
    },
    geometry: {
      type: 'Point',
      coordinates: [1923, 2093]
    }
  }, {
    type: 'Feature',
    properties: {
      name: 'Hispania'
    },
    geometry: {
      type: 'Point',
      coordinates: [1374, 2148]
    }
  }]

  // geoJson definitions
  window.geoInfo = [{
    type: 'Feature',
    properties: {
      name: 'Mare Germanicum'
    },
    geometry: {
      type: 'Point',
      coordinates: [1589, 1447]
    }
  }, {
    type: 'Feature',
    properties: {
      name: 'Mare Balticum'
    },
    geometry: {
      type: 'Point',
      coordinates: [2090, 1407]
    }
  }, {
    type: 'Feature',
    properties: {
      name: 'Mare Mediteraneum'
    },
    geometry: {
      type: 'Point',
      coordinates: [2028, 2453]
    }
  }, {
    type: 'Feature',
    properties: {
      name: 'Mare Maggiore'
    },
    geometry: {
      type: 'Point',
      coordinates: [2623, 1918]
    }
  }]

  // polygon
  window.polygon = [{
    x: 1528.5,
    y: 1524
  }, {
    x: 1532,
    y: 1571
  }, {
    x: 1559.5,
    y: 1620.5
  }, {
    x: 1541,
    y: 1612
  }, {
    x: 1562.5,
    y: 1634.5
  }, {
    x: 1548.5,
    y: 1655
  }, {
    x: 1567.5,
    y: 1651
  }, {
    x: 1598.5,
    y: 1666.5
  }, {
    x: 1576,
    y: 1713
  }, {
    x: 1562,
    y: 1718
  }, {
    x: 1584,
    y: 1722.5
  }, {
    x: 1586.5,
    y: 1736.5
  }, {
    x: 1550.5,
    y: 1750.5
  }, {
    x: 1505,
    y: 1743.5
  }, {
    x: 1505.5,
    y: 1752
  }, {
    x: 1493.5,
    y: 1754.5
  }, {
    x: 1480,
    y: 1745.5
  }, {
    x: 1462,
    y: 1742
  }, {
    x: 1448.5,
    y: 1733
  }, {
    x: 1431.5,
    y: 1737.5
  }, {
    x: 1421,
    y: 1755
  }, {
    x: 1403,
    y: 1740.5
  }, {
    x: 1377.5,
    y: 1741.5
  }, {
    x: 1370.5,
    y: 1751
  }, {
    x: 1349.5,
    y: 1733
  }, {
    x: 1367.5,
    y: 1725
  }, {
    x: 1400,
    y: 1715
  }, {
    x: 1417.5,
    y: 1704.5
  }, {
    x: 1445,
    y: 1709
  }, {
    x: 1463,
    y: 1696
  }, {
    x: 1436.5,
    y: 1703
  }, {
    x: 1427,
    y: 1690
  }, {
    x: 1415,
    y: 1679
  }, {
    x: 1397.5,
    y: 1683
  }, {
    x: 1386.5,
    y: 1667.5
  }, {
    x: 1424.5,
    y: 1655
  }, {
    x: 1428,
    y: 1632
  }, {
    x: 1406.5,
    y: 1636
  }, {
    x: 1435.5,
    y: 1613
  }, {
    x: 1457,
    y: 1614
  }, {
    x: 1467,
    y: 1593.5
  }, {
    x: 1457.5,
    y: 1572.5
  }, {
    x: 1467.5,
    y: 1551
  }, {
    x: 1528.5,
    y: 1524
  }]
}(window))


example/index.js
/*
 * @copyright 2015 commenthol
 * @license MIT
 */

/* global L */

;(function (window) {
  function init (mapid) {
    const minZoom = 0
    const maxZoom = 5
    const img = [
      3831, // original width of image `karta.jpg`
      3101 // original height of image
    ]

    // create the map
    const map = L.map(mapid, {
      crs: L.CRS.Simple,
      minZoom,
      maxZoom
    })

    // assign map and image dimensions
    const rc = new L.RasterCoords(map, img)

    // set the view on a marker ...
    map.setView(rc.unproject([1589, 1447]), 4)

    // add layer control object
    L.control.layers({}, {
      Polygon: layerPolygon(map, rc),
      Countries: layerCountries(map, rc),
      Bounds: layerBounds(map, rc, img),
      Info: layerGeo(map, rc),
      Circles: layerCircles(map, rc)
    }).addTo(map)

    // the tile layer containing the image generated with gdal2tiles --leaflet ...
    L.tileLayer('./tiles/{z}/{x}/{y}.png', {
      noWrap: true,
      bounds: rc.getMaxBounds(),
      maxNativeZoom: rc.zoomLevel(),
      attribution: 'Map <a href="https://commons.wikimedia.org/wiki/' +
        'File:Karta_%C3%B6ver_Europa,_1672_-_Skoklosters_slott_-_95177.tif">' +
        'Karta över Europa, 1672 - Skoklosters</a> under ' +
        '<a href="https://creativecommons.org/publicdomain/zero/1.0/deed.en">CC0</a>'
    }).addTo(map)
  }

  /**
   * layer with markers
   */
  function layerBounds (map, rc, img) {
    // set marker at the image bound edges
    const layerBounds = L.layerGroup([
      L.marker(rc.unproject([0, 0])).bindPopup('[0,0]'),
      L.marker(rc.unproject(img)).bindPopup(JSON.stringify(img))
    ])
    map.addLayer(layerBounds)

    // set markers on click events in the map
    map.on('click', function (event) {
      // to obtain raster coordinates from the map use `project`
      const coord = rc.project(event.latlng)
      // to set a marker, ... in raster coordinates in the map use `unproject`
      const marker = L.marker(rc.unproject(coord))
        .addTo(layerBounds)
      marker.bindPopup('[' + Math.floor(coord.x) + ',' + Math.floor(coord.y) + ']')
        .openPopup()
    })

    return layerBounds
  }

  /**
   * layer using geoJson data for countries adding a circle marker
   */
  function layerCountries (map, rc) {
    const layerCountries = L.geoJson(window.countries, {
      // correctly map the geojson coordinates on the image
      coordsToLatLng: function (coords) {
        return rc.unproject(coords)
      },
      // add a popup content to the marker
      onEachFeature: function (feature, layer) {
        if (feature.properties && feature.properties.name) {
          layer.bindPopup(feature.properties.name)
        }
      },
      pointToLayer: function (feature, latlng) {
        return L.circleMarker(latlng, {
          radius: 8,
          fillColor: '#800080',
          color: '#D107D1',
          weight: 1,
          opacity: 1,
          fillOpacity: 0.8
        })
      }
    })
    map.addLayer(layerCountries)
    return layerCountries
  }

  /**
   * layer with red markers
   */
  function layerGeo (map, rc) {
    const imgDir = 'images/'
    const redMarker = L.icon({
      iconUrl: imgDir + 'marker-icon-red.png',
      iconRetinaUrl: imgDir + 'marker-icon-red-2x.png',
      iconSize: [25, 41],
      iconAnchor: [12, 41],
      popupAnchor: [-0, -31],
      shadowUrl: imgDir + 'marker-shadow.png',
      shadowSize: [41, 41],
      shadowAnchor: [14, 41]
    })
    const layerGeo = L.geoJson(window.geoInfo, {
      // correctly map the geojson coordinates on the image
      coordsToLatLng: function (coords) {
        return rc.unproject(coords)
      },
      // add a popup content to the marker
      onEachFeature: function (feature, layer) {
        if (feature.properties && feature.properties.name) {
          layer.bindPopup(feature.properties.name)
        }
      },
      pointToLayer: function (feature, latlng) {
        return L.marker(latlng, {
          icon: redMarker
        })
      }
    })
    map.addLayer(layerGeo)
    return layerGeo
  }

  /**
   * layer drawing a polygon
   */
  function layerPolygon (map, rc) {
    const points = window.polygon.map(function (point) {
      return rc.unproject([point.x, point.y])
    })
    const layerPolygon = L.polygon([points])
    map.addLayer(layerPolygon)
    return layerPolygon
  }

  /**
   * layer drawing some cicles
   */
  function layerCircles (map, rc) {
    /*
    // using circle may cause displaying a ellipse at the edges of the image
    // radius is painful to adjust - simply don't use
    const circle = L.circle(rc.unproject([200, 1000]), { radius: 1e6 })
    */

    /*
    // drawing a circle with a polyline
    // Not so nice because of the visible steps
    function circlePoints ([x, y], r, steps = 360) {
      var p = []
      for (var i = 0; i < steps; i++) {
        p.push(rc.unproject([
          (x + r * Math.cos(2 * Math.PI * i / steps)),
          (y + r * Math.sin(2 * Math.PI * i / steps))
        ]))
      }
      return p
    }
    const polyline = L.polygon([circlePoints([200, 200], 200)], {
      fillColor: '#3388ff',
      color: '#fb0000'
    })
    */

    // Custom marker prototype - credits to Arkensor
    L.CircleMarkerScaling = L.CircleMarker.extend({
      _project: function () {
        this._point = this._map.latLngToLayerPoint(this._latlng)
        this._radius = 2 * this.options.radius * this._map.getZoomScale(this._map.getZoom(), this._map.getMaxZoom())
        this._updateBounds()
      }
    })
    L.circleMarkerScaling = function (latlng, options) {
      return new L.CircleMarkerScaling(latlng, options)
    }

    const custom = L.circleMarkerScaling(rc.unproject([200, 200]), {
      radius: 200,
      fillColor: '#3388ff',
      color: '#fbff2c'
    })

    const layer = L.featureGroup([custom])
    map.addLayer(layer)
    return layer
  }

  init('map')
}(window))
